runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/resources/PipelineResources.java

```diff
@@ -64,6 +70,31 @@
     return resourcePath -> !resourcePath.contains("gradle/wrapper");
   }
 
+  /**
+   * Goes through the list of files that need to be staged on runner. Removes nonexistent
+   * directories and packages existing ones. This is necessary for runners that require filesToStage
+   * to be jars only.
+   *
+   * @param options options object with the files to stage and temp location for staging
+   */
+  public static void prepareFilesForStaging(FileStagingOptions options) {
+    List<String> filesToStage = options.getFilesToStage();
+    if (filesToStage == null || filesToStage.isEmpty()) {
+      filesToStage = detectClassPathResourcesToStage(ReflectHelpers.findClassLoader(), options);
```

Comment:
I am assuming that this is the 'correct' way to find the correct classloader, it is the only possible breaking change of this PR

Reply:
I did have a problem the context classloader (which is the first choice for `ReflectHelpers.findClassLoader()`) in the Spark runner before - BEAM-10109. But the class loader should only be mutated like that when pipeline construction and Spark execution are happening in the same JVM, so as long as we don't rely on  `prepareFilesForStaging` for tests I don't anticipate it being a problem.

Reply:
Thanks for the pointer, I am going to try this at least on YARN to be cautious about any possible deployment issue.